Update and improve video capture sample

ngphibang opened the pull request:
As mt9m114 and ov5640 camera shields have been merged. The sample needs to be updated too.
Some improvements on the video capture sample are also made:

- Display frames on LCD panel instead of just discarding them
- Add support for **RT1170-evk** and video sw generator

This PR is split from https://github.com/zephyrproject-rtos/zephyr/pull/69810 to ease the review process as required by @loicpoulain.
~~This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/72434 and https://github.com/zephyrproject-rtos/zephyr/pull/72435~~

ngphibang committed:
samples: video: capture: Use a chosen node for camera

In order to be generic, use a chosen node for camera so that the sample
is not specific to NXP SoCs. Also, always favorite a real video device
unless it is unavailable.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Display captured frames on screen

Improve the sample application by displaying the captured frames instead
of just discarding them.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Run clang-format

Run clang-format on the file before making any changes

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Add support for i.MX RT1170 EVK

Add support for i.MX RT1170 EVK

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented:
Nice to see some complete device-to-device workflow.

This looks fairly similar to this pull request: https://github.com/zephyrproject-rtos/zephyr/pull/71463
Would it make more sense to split the "capture-only" or "capture to display" like this other pull request does?

ngphibang commented:
Thanks, I am not aware of the PR you pointed out. BTW, this PR is part of the series https://github.com/zephyrproject-rtos/zephyr/pull/71463 which I did last year and first pushed on March 5, 2024.

About the "capture-only" vs "captue-to-display", I think

- It does not make much sense if we only capture frames and discard them. The original video capture sample normally should display the frame, but I think for some reasons, it was not done at that time.
-This PR itself allows capture-only or capture to display depending on the availability of the hardwares.

josuah commented:
Thank you for explaining, I was curious... The other example uses LVGL: not the same strategy.

ngphibang commented:
@mmahadevan108 : yes, it's tested on 1064_evk, 1170_evk and evkb.

loicpoulain commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -104,6 +138,16 @@ int main(void)
 		printk("\rGot frame %u! size: %u; timestamp %u ms\n", frame++, vbuf->bytesused,
 		       vbuf->timestamp);
 
+#if DT_HAS_CHOSEN(zephyr_display)
```
Could you please make the content a function (e.g `video_display_frame()`) ? (and no-op if !zephyr_display)

ngphibang replied:
yes, done. I also made a display_setup() function. Thanks

danieldegrasse commented on samples/subsys/video/capture/prj.conf:
```diff
@@ -4,3 +4,4 @@ CONFIG_SHELL=y
 CONFIG_DEVICE_SHELL=y
 CONFIG_PRINTK=y
 CONFIG_LOG=y
+CONFIG_DISPLAY=y
```
Should we enable `CONFIG_DISPLAY` by default in the sample? If we do, we should insure that devices with a display node that is not enabled (IE `status = "disabled"`) still build correctly, as do devices with no display node defined.

ngphibang replied:
`CONFIG_DISPLAY` is a generic option just like 'CONFIG_VIDEO'. I tested and see it build with or without `zephyr_display `node defined, so no problem here.

The only thing to care is when `zephyr_display `is defined, its corresponding display node (here eLCDIF) needs to be enabled as well. If not, it will not build (despite of `CONFIG_DISPLAY` enabled or not)

danieldegrasse commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -57,43 +53,72 @@ int main(void)
+	       capabilities.supported_pixel_formats, capabilities.current_pixel_format,
+	       capabilities.current_orientation);
+
+#if defined(CONFIG_VIDEO_MCUX_MIPI_CSI2RX)
```
I think we should try to avoid interspersing device specific Kconfigs within the sample as much as possible- Perhaps we could define a few constants at the top of this file based on which video devices were enabled, like so?

```
#ifdef CONFIG_VIDEO_MCUX_MIPI_CSI2RX
#define DISPLAY_FORMAT PIXEL_FORMAT_ARGB_8888
#else
#define DISPLAY_FORMAT PIXEL_FORMAT_RGB565
#endif

#ifdef CONFIG_MCUX_ELCID_PXP
#define BUFF_ALIGN 64
#else
#define BUFF_ALIGN 1
#endif
```

My reasoning here is that this allows us to keep all the driver-specific KConfigs in one location within the sample, so that snippets like these don't get interspersed within the sample to handle specific hardware requirements

ngphibang replied:
Yes, done. Just a small remark, when there is no alignment required, it is actually `sizeof(void *)` not 1 :-).

kartben commented:
cc @CharlesDias

ngphibang commented:
CI failures are not due to the PR changes.

danieldegrasse commented on samples/subsys/video/capture/boards/nxp_rt11xx.conf:
I think that duplicating configuration settings here is worth the tradeoff. Personally I find it confusing when testing a sample requires additional build flags on the command line (like you would in this case to pull in the RT11XX overlay). I think we should instead enable these options in board specific overlays, so that customers will simply be able to build the sample for the given board and provided their hardware is setup correctly everything will "just work"

ngphibang committed:
samples: video: capture: Update documents and tests

The mt9m114 camera shield is now added. There are also some changes in
the mt9m114 camera driver, e.g. frame rate, default format, capabilities.
Update the sample document and test to reflect these changes.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Enhance logs readability

There are some compatibilty situations where carriage return does not
work (e.g. on Serial Monitor in VSCode). Moreover, keeping the
timestamps logs on the console would help to have an idea about the
frame rate. So, it's better to use line feed instead of carriage return
in this case.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
- Updated the documentation to use a display shield for RT1064 because display stuffs WILL be removed from RT1064 EVK by [this PR](https://github.com/zephyrproject-rtos/zephyr/pull/72390) (which should be merged first)

- Some reorganisation of the commits.

@loicpoulain @danieldegrasse Could you revisit this ?

josuah commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -15,39 +16,85 @@ LOG_MODULE_REGISTER(main);
+		LOG_ERR("%s: video device is not ready", video_dev->name);
 		return 0;
 	}
+#else
```
Is the `#else` really required? It looks like the device name is a built-time constant, so resolving the name at runtime with `device_get_binding()` is not required?
I am still unsure about which one to choose in which context.
An alternative would be adding an empty `#else` block with `#error` to hint the user about using a chosen block.

ngphibang replied:
`#else` is required because if we don't have a real video device hardware, we use the software generator, but if we have one, we use the hardware, **not** both.

josuah reviewed (approved):
Glad to see all this past work making it all the way to the user-facing samples!

ngphibang commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -15,39 +16,85 @@ LOG_MODULE_REGISTER(main);
+#define DISPLAY_FORMAT PIXEL_FORMAT_BGR_565
+#endif
+
+#ifdef CONFIG_MCUX_ELCDIF_PXP
+#define BUFFER_ALIGN 64
+#endif
```
yes, it's better with a Kconfig option _in case that_ the hardware (PxP) really requires buffer alignement to work, and then driver will check it in code. But here, aligment is **not** mandatory though PxP performance will be improved if the buffer is (64B) aligned.

ngphibang replied:
@josuah : Finally, I found and used the `CONFIG_VIDEO_BUFFER_POOL_ALIGN` which is available in the video subsystem (it is there but no one uses it :) ). This way, we can avoid platform specific stuffs in the code.

josuah replied:
Great news! Putting it in the example also brings some context and visibility for it. Two birds one stone.

danieldegrasse commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -106,7 +106,7 @@ int main(void)
 		}
 
-		printk("\rGot frame %u! size: %u; timestamp %u ms", frame++, vbuf->bytesused,
+		printk("\rGot frame %u! size: %u; timestamp %u ms\n", frame++, vbuf->bytesused,
```
Hmmm. I think based on the `\r` at the start of this print statement, the choice to replace the current line with a new one was intentional. I'm ok to make this change, but we should drop the `\r` if we are changing this

ngphibang replied:
Yes, for me, `\r` is working on Teraterm but not on Serial Monitor of VSCode. "\n" is working for both platform. And by keeping the logs (with \n), we can look at the timestamps to compute the frame rate which to me, is better.

BTW, as you said, \r should be removed.

danieldegrasse commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -15,6 +16,48 @@ LOG_MODULE_REGISTER(main);
+
+	display_write(display_dev, 0, 0, &buf_desc, vbuf->buffer);
+
+	display_blanking_off(display_dev);
```
Why do we need to call `display_blanking_off` every time here? Can't we just call it once at init?

ngphibang replied:
Good remark. I will look at this. Thanks

danieldegrasse commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -15,6 +16,48 @@ LOG_MODULE_REGISTER(main);
+	       capabilities.supported_pixel_formats, capabilities.current_pixel_format,
+	       capabilities.current_orientation);
+
+	display_set_pixel_format(display_dev, DISPLAY_FORMAT);
```
What is the reasoning behind setting the format here? Some displays will likely want to use a different format than the one this sample sets- it is really that we need the camera format and display format to match, right?

I suppose maybe what we need to do is only enable the display if the camera format is RGB565? I suppose we could also enable XRGB888 via the ARGB8888 display format

ngphibang replied:
> What is the reasoning behind setting the format here? Some displays will likely want to use a different format than the one this sample sets- it is really that we need the camera format and display format to match, right?

Yes, the camera and display formats must match. Normally, it is the (human) user who needs to look at both camera and display default formats (and capabilities) then sets the format. But here, we want the sample work "out-of-the-box", for RT1170 case, the camera output format is XRGB so we need to set the display (we have many displays but here in this sample, it is rk055hdmi4ma0) to this format.

> I suppose maybe what we need to do is only enable the display if the camera format is RGB565? 

It's a big pity for me if we do that. The camera output on RT1170 is XRGB8888 and we cannot have camera frame displayed on LCD just because of the mismatch formats between camera and display ? I know that you want to be generic but the sample supports RT1064 and RT1170 as reference boards, we can always set the display format to match these cases, no ?

> I suppose we could also enable XRGB888 via the ARGB8888 display format

Yes, XRGB is more generic than ARGB and we can use both alternatively. But because there is no PIXEL_FORMAT_XRGB_8888 defined in display.h, I used the other.

danieldegrasse replied:
> I think it will be a big pity if we do that. The camera output on RT1170 is XRGB8888 and we cannot have camera frames displayed on LCD just because of the mismatch of the default formats between camera and display ? I know that you want to be generic but the sample supports RT1064 and RT1170 as reference boards, we can always set the display format to match these cases, no ? Also, through this, it will give some hints to people that we could "always" match the camera / display formats for their own camera / display set-up.

I think I'm missing something here. I guess the PXP will convert the XRGB8888 data to RGB565 for the LCD? *Ideally*, the display pipeline would report its format at XRGB8888 in this case. From my understanding, the application either requires that the format of the buffer that is output from the video pipeline and input to the display pipeline be the same, or the application converts the buffer to a new format manually.

ngphibang replied:
> I think I'm missing something here. I guess the PXP will convert the XRGB8888 data to RGB565 for the LCD ?

No, as you can see, the current PxP driver does nothing apart from rotation. It's the eLCDIF who can support various formats (via my PR which was merged sometimes ago).

> From my understanding, the application either requires that the format of the buffer that is output from the video pipeline and input to the display pipeline be the same,

I don't think so, as you can see, display controller (like eLCDIF) can support various formats, so application can set the display format. It is the actual situation now.

> or the application converts the buffer to a new format manually.

Yes, this is another way. Application can use software or hardware like PxP to do so (I am working on making a new PxP driver to do this).

danieldegrasse replied:
> No, as you can see, the current PxP driver does nothing apart from rotation. It's the eLCDIF itself who can support XRGB format (via https://github.com/zephyrproject-rtos/zephyr/pull/69250 which was merged sometimes ago).

Sure- sorry, I remembered seeing something for this, but wasn't sure where we added the support.

> I don't think so, as you can see, display controller (like eLCDIF) can support various formats, so application can set the display format at runtime, it does not require that the camera output format is the same as the default display format. It is the actual situation.

Yes, I follow that. To be clear, what I am asking for here is the following code:

```
/* Set display pixel format to match the one in use by the camera */
switch (capabilities.current_pixel_format) {
case VIDEO_PIX_FMT_RGB565:
    display_set_pixel_format(display_dev, PIXEL_FORMAT_BGR_565);
case VIDEO_PIX_FMT_XRGB32:
    display_set_pixel_format(display_dev, PIXEL_FORMAT_ARGB_8888);
default:
    return -ENOTSUP;
}
```

My understanding is that the Sensor->CSI pipeline on the RT1064 would output `VIDEO_PIX_FMT_RGB565` as the current format, and the Sensor->CSI2RX->CSI pipeline on the RT1170 would output `VIDEO_PIX_FMT_XRGB32` as the current format. Is that incorrect?

ngphibang replied:
Yes, this is definitely more elegant :). Thanks, done.

ngphibang commented:
CI twister test failure on RT1064 is because the` zephyr,display` chosen node is left and the node is disabled when display stuffs are already moved to a shield by [this PR](https://github.com/zephyrproject-rtos/zephyr/pull/72390)

danieldegrasse reviewed (approved):
One minor nit- otherwise lgtm. Going to approve given the time constraints of the merge window

ngphibang commented on samples/subsys/video/capture/src/main.c:
```diff
@@ -37,6 +38,23 @@ static inline int display_setup(const struct device *const display_dev)
 	       capabilities.supported_pixel_formats, capabilities.current_pixel_format,
 	       capabilities.current_orientation);
 
+	/* Set display pixel format to match the one in use by the camera */
```
I thought about this. I think both ways are ok. The previous commit adds support for display for the current supported board RT1064 which actually works without the needs of this change because the default format of camera and display already matches. This change is needed essentially when we add support for RT1170 EVK.
